### ISSUE TITLE
CQL to ELM translator, date range optimization hints are not returned for sub-queries

### ISSUE THREAD

**Issue by amarvakul** [cqframework/clinical_quality_language#545]
CQL to ELM translator, date range optimization hints are not returned for sub-queries

CQL

`define "Claim Retrieve1": [Claim] C where exists C.item CI where (CI.serviced intersect "Default Claim Period") is not null
`

ELM 

> <def name="Claim Retrieve1" context="Patient" accessLevel="Public">
         <resultTypeSpecifier xsi:type="ListTypeSpecifier">
            <elementType name="fhir:Claim" xsi:type="NamedTypeSpecifier"/>
         </resultTypeSpecifier>
         <expression xsi:type="Query">
            <resultTypeSpecifier xsi:type="ListTypeSpecifier">
               <elementType name="fhir:Claim" xsi:type="NamedTypeSpecifier"/>
            </resultTypeSpecifier>
            <source alias="C">
               <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                  <elementType name="fhir:Claim" xsi:type="NamedTypeSpecifier"/>
               </resultTypeSpecifier>
               <expression dataType="fhir:Claim" xsi:type="Retrieve">
                  <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                     <elementType name="fhir:Claim" xsi:type="NamedTypeSpecifier"/>
                  </resultTypeSpecifier>
               </expression>
            </source>
            <where resultTypeName="t:Boolean" xsi:type="Exists">
               <signature xsi:type="ListTypeSpecifier">
                  <elementType name="fhir:Claim.Item" xsi:type="NamedTypeSpecifier"/>
               </signature>
               <operand xsi:type="Query">
                  <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                     <elementType name="fhir:Claim.Item" xsi:type="NamedTypeSpecifier"/>
                  </resultTypeSpecifier>
                  <source alias="CI">
                     <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                        <elementType name="fhir:Claim.Item" xsi:type="NamedTypeSpecifier"/>
                     </resultTypeSpecifier>
                     <expression path="item" scope="C" xsi:type="Property">
                        <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                           <elementType name="fhir:Claim.Item" xsi:type="NamedTypeSpecifier"/>
                        </resultTypeSpecifier>
                     </expression>
                  </source>
                  <where resultTypeName="t:Boolean" xsi:type="Not">
                     <signature name="t:Boolean" xsi:type="NamedTypeSpecifier"/>
                     <operand resultTypeName="t:Boolean" xsi:type="IsNull">
                        <signature name="t:Any" xsi:type="NamedTypeSpecifier"/>
                        <operand xsi:type="Intersect">
                           <resultTypeSpecifier xsi:type="IntervalTypeSpecifier">
                              <pointType name="t:DateTime" xsi:type="NamedTypeSpecifier"/>
                           </resultTypeSpecifier>
                           <signature xsi:type="IntervalTypeSpecifier">
                              <pointType name="t:DateTime" xsi:type="NamedTypeSpecifier"/>
                           </signature>
                           <signature xsi:type="IntervalTypeSpecifier">
                              <pointType name="t:DateTime" xsi:type="NamedTypeSpecifier"/>
                           </signature>
                           <operand name="ToInterval" libraryName="FHIRHelpers" xsi:type="FunctionRef">
                              <signature name="fhir:Period" xsi:type="NamedTypeSpecifier"/>
                              <operand asType="fhir:Period" xsi:type="As">
                                 <operand path="serviced" scope="CI" xsi:type="Property">
                                    <resultTypeSpecifier xsi:type="ChoiceTypeSpecifier">
                                       <choice name="fhir:date" xsi:type="NamedTypeSpecifier"/>
                                       <choice name="fhir:Period" xsi:type="NamedTypeSpecifier"/>
                                    </resultTypeSpecifier>
                                 </operand>
                              </operand>
                           </operand>
                           <operand name="Default Claim Period" xsi:type="ExpressionRef">
                              <resultTypeSpecifier xsi:type="IntervalTypeSpecifier">
                                 <pointType name="t:DateTime" xsi:type="NamedTypeSpecifier"/>
                              </resultTypeSpecifier>
                           </operand>
                        </operand>
                     </operand>
                  </where>
               </operand>
            </where>
         </expression>
      </def>

Another example where date range optimization is not returned when a "date" to "date" type comparison is made

CQL
`define DateOnlyCondition: [Condition] C where C.assertedDate < start of "Measurement Period"`

ELM
```
<def name="DateOnlyCondition" context="Patient" accessLevel="Public">
         <resultTypeSpecifier xsi:type="ListTypeSpecifier">
            <elementType name="fhir:Condition" xsi:type="NamedTypeSpecifier"/>
         </resultTypeSpecifier>
         <expression xsi:type="Query">
            <resultTypeSpecifier xsi:type="ListTypeSpecifier">
               <elementType name="fhir:Condition" xsi:type="NamedTypeSpecifier"/>
            </resultTypeSpecifier>
            <source alias="C">
               <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                  <elementType name="fhir:Condition" xsi:type="NamedTypeSpecifier"/>
               </resultTypeSpecifier>
               <expression dataType="fhir:Condition" xsi:type="Retrieve">
                  <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                     <elementType name="fhir:Condition" xsi:type="NamedTypeSpecifier"/>
                  </resultTypeSpecifier>
               </expression>
            </source>
            <where resultTypeName="t:Boolean" xsi:type="Less">
               <signature name="t:DateTime" xsi:type="NamedTypeSpecifier"/>
               <signature name="t:DateTime" xsi:type="NamedTypeSpecifier"/>
               <operand name="ToDateTime" libraryName="FHIRHelpers" xsi:type="FunctionRef">
                  <signature name="fhir:dateTime" xsi:type="NamedTypeSpecifier"/>
                  <operand resultTypeName="fhir:dateTime" path="assertedDate" scope="C" xsi:type="Property"/>
               </operand>
               <operand resultTypeName="t:DateTime" xsi:type="Start">
                  <signature xsi:type="IntervalTypeSpecifier">
                     <pointType name="t:DateTime" xsi:type="NamedTypeSpecifier"/>
                  </signature>
                  <operand name="Measurement Period" xsi:type="ParameterRef">
                     <resultTypeSpecifier xsi:type="IntervalTypeSpecifier">
                        <pointType name="t:DateTime" xsi:type="NamedTypeSpecifier"/>
                     </resultTypeSpecifier>
                  </operand>
               </operand>
            </where>
         </expression>
      </def>
```

Expectations: "dateRange" and "dateProperty" to be returned either at the top level query or in a completely separate structure outside of the  "statements" in the ELM. As we are parsing these optimization hints and push them as predicates at the time of data access of a resource otherwise we would not have these hints information at the right if we only find them at the sub-query.

**Comment by brynrhodes** [comment]
Note that this is currently being worked on as a $datarequirements evaluation component in the CQF Tooling repository.

**Comment by brynrhodes** [comment]
This is addressed by the new DataRequirementsProcessor, included in the 1.5.4 release, in that data requirements returned from that analysis will include date range optimizations.
